src/NUnitFramework/framework/Internal/Results/TestSuiteResult.cs

```diff
@@ -112,46 +117,53 @@ public override IList<ITestResult> Children
         /// <param name="result">The result to be added</param>
         public virtual void AddResult(ITestResult result)
         {
-            Children.Add(result);
+            Children.Enqueue(result);
 
             //AssertCount += result.AssertCount;
 
-            // If this result is marked cancelled, don't change it
-            if (ResultState != ResultState.Cancelled)
+            bool stateSet;
+            do
+            {
+                stateSet = true;
+
+                var resultState = ResultState;
+
+                // If this result is marked cancelled, don't change it
+                if (resultState == ResultState.Cancelled)
+                    break;
+
                 switch (result.ResultState.Status)
                 {
                     case TestStatus.Passed:
 
-                        if (ResultState.Status == TestStatus.Inconclusive)
-                            SetResult(ResultState.Success);
+                        if (resultState.Status == TestStatus.Inconclusive)
+                            stateSet = SetResultIf(resultState, ResultState.Success);
 
                         break;
 
                     case TestStatus.Failed:
 
 
-                        if (ResultState.Status != TestStatus.Failed)
-                            SetResult(ResultState.ChildFailure, CHILD_ERRORS_MESSAGE);
+                        if (resultState.Status != TestStatus.Failed)
+                            stateSet = SetResultIf(resultState, ResultState.ChildFailure, CHILD_ERRORS_MESSAGE);
 
                         break;
 
                     case TestStatus.Skipped:
 
                         if (result.ResultState.Label == "Ignored")
-                            if (ResultState.Status == TestStatus.Inconclusive || ResultState.Status == TestStatus.Passed)
-                                SetResult(ResultState.Ignored, CHILD_IGNORE_MESSAGE);
-
-                        break;
+                            if (resultState.Status == TestStatus.Inconclusive || resultState.Status == TestStatus.Passed)
+                                stateSet = SetResultIf(resultState, ResultState.Ignored, CHILD_IGNORE_MESSAGE);
 
-                    default:
                         break;
                 }
+            } while (!stateSet);
 
-            AssertCount += result.AssertCount;
-            _passCount += result.PassCount;
-            _failCount += result.FailCount;
-            _skipCount += result.SkipCount;
-            _inconclusiveCount += result.InconclusiveCount;
+            Interlocked.Add (ref InternalAssertCount, result.AssertCount);
+            Interlocked.Add (ref _passCount, result.PassCount);
+            Interlocked.Add (ref _failCount, result.FailCount);
+            Interlocked.Add (ref _skipCount, result.SkipCount);
+            Interlocked.Add (ref _inconclusiveCount, result.InconclusiveCount);
         }
 
```

Comment:
What I mean is that we may have (for example) a passing fixture that contains a failing test for a brief time. That's an invalid state. We have these invalid transitional states in objects all the time and they (usually) cause no problem. The question is whether it causes a problem here, with multiple threads potentially accessing the parent result.

As a thought experiment, I'm imagining a fixture result that has so far succeeded. Now two tests run in parallel. One of them fails and the other is ignored. The parent result should end up as a failure, no matter how the two different test thread executions are interleaved. 

As a worst case, assume that the two threads are both reaching step 2 at approximately the same time. Both of them discover that the parent result is passing. They are competing to change that result - in one case to failure in the other to ignrored. As I read the code, if the ignored test gets there first, the result will be ignored. The failing test will discover that the result is changed and try again, finally making the result a failure. If the failing test gets there first, it will be the ignored test that tries again and it will back off not needing to change anything.

Is that how you see the particular scenario playing out as well?